cinder/backup/chunkeddriver.py

```diff
@@ -780,3 +776,161 @@ def delete_backup(self, backup):
                 eventlet.sleep(0)
 
         LOG.debug('delete %s finished.', backup['id'])
+
+
+class BackupRestoreHandle(object):
+    """Class used to reconstruct a backup from chunks."""
+    def __init__(self, chunked_driver):
+        self._driver = chunked_driver
+        self._segments = []
+        self._object_readers = {}
+        self._idx = -1
+
+    def has_next(self):
+        return self._idx < len(self._segments) - 1
+
+    def get_next(self):
+        self._idx += 1
+        return self._segments[self._idx]
+
+    def read(self, segment):
+        """Reads the bytes of a segment"""
+        buff_reader = self._get_reader(segment)
+        # seek inside the backup chunk containing this segment
+        offset_diff = segment.offset - segment.obj['offset']
+        buff_reader.seek(offset_diff)
+        # read the segment's length from the chunk
+        data = buff_reader.read(segment.length)
+        eventlet.tpool.execute(self._clear_reader, segment)
+        return data
+
+    def _get_reader(self, segment):
+        """Keeps an internal cache of object readers.
+
+        Avoids calling the storage backend multiple times for objects which
+        have been split into multiple segments due to merging.
+        """
+        obj_name = segment.obj['name']
+        obj_reader = self._object_readers.get(obj_name)
+        if not obj_reader:
+            obj_reader = self._get_new_reader(segment)
+            self._object_readers[obj_name] = obj_reader
+        return obj_reader
+
+    def _get_new_reader(self, segment):
+        with self._driver._get_object_reader(
+                segment.obj['container'],
+                segment.obj['name'],
+                extra_metadata=segment.obj['extra_metadata']) \
+                as reader:
+            return io.BytesIO(self._get_raw_bytes(reader, segment.obj))
+
+    def _get_raw_bytes(self, reader, obj):
+        """Get the bytes of a backup chunk, decompressing if needed"""
+        compression_algorithm = obj['compression']
+        decompressor = self._driver._get_compressor(compression_algorithm)
+        if decompressor is not None:
+            LOG.debug('decompressing data using %s algorithm',
+                      compression_algorithm)
+            return decompressor.decompress(reader.read())
+        return reader.read()
+
+    def _clear_reader(self, segment):
+        """Clear the object reader for a segment, if needed.
+
+        If there is no further segment for the same object, we close and
+        remove the corresponding object reader, freeing up the memory.
+        """
+        obj_name = segment.obj['name']
+        for _segment in self._segments[self._idx:]:
+            if obj_name == _segment.obj['name']:
+                return
+
+        obj_reader = self._object_readers[obj_name]
```

Comment:
My bad.
And looking at it again, I would even remove the `if` and make it
````
        self._object_readers[obj_name].close()
        self._object_readers.pop(obj_name)
````
because there's only a single point when the reader is being freed - when we're done reading all the segments of the same object.